File: backend/internal/scheduler/fcfs.go
package scheduler

import (
	"sort"

	"github.com/pdro1812/trabalho-os-scheduler/backend/internal/models"
)

func RunFCFS(req models.SimulationRequest) models.SimulationResponse {
	processes := make([]models.Process, len(req.Processes))
	copy(processes, req.Processes)

	// Ordena por Arrival Time (Garante o FCFS caso tenham tempos de chegada diferentes)
	sort.SliceStable(processes, func(i, j int) bool {
		return processes[i].ArrivalTime < processes[j].ArrivalTime
	})

	var executionOrder []models.ExecutionBlock
	var metrics []models.ProcessMetrics

	currentTime := 0
	totalWaitTime := 0
	totalTurnaroundTime := 0

	for i, p := range processes {
		// Se a CPU estiver ociosa esperando o processo chegar
		if currentTime < p.ArrivalTime {
			currentTime = p.ArrivalTime
		}

		start := currentTime
		currentTime += p.BurstTime
		end := currentTime

		executionOrder = append(executionOrder, models.ExecutionBlock{
			PID:   p.PID,
			Start: start,
			End:   end,
		})

		turnaround := end - p.ArrivalTime
		wait := turnaround - p.BurstTime

		metrics = append(metrics, models.ProcessMetrics{
			PID:           p.PID,
			EffectiveTime: turnaround,
			WaitingTime:   wait,
		})

		totalTurnaroundTime += turnaround
		totalWaitTime += wait

		// Aplica TTC se não for o último processo a executar
		if i < len(processes)-1 && req.TTC > 0 {
			executionOrder = append(executionOrder, models.ExecutionBlock{
				PID:   "TTC",
				Start: currentTime,
				End:   currentTime + req.TTC,
			})
			currentTime += req.TTC
		}
	}

	n := float64(len(processes))
	return models.SimulationResponse{
		ExecutionOrder:    executionOrder,
		Metrics:           metrics,
		AvgWaitTime:       float64(totalWaitTime) / n,
		AvgTurnaroundTime: float64(totalTurnaroundTime) / n,
	}
}

File: backend/internal/scheduler/fcfs.js
function runFCFS(requisicao) {
    // Pegamos a lista de processos da requisição
    const processos = [];
    for (let i = 0; i < requisicao.Processes.length; i++) {
        processos.push(requisicao.Processes[i]);
    }

    // Primeiro, ordenamos os processos por tempo de chegada (ArrivalTime)
    processos.sort((a, b) => a.ArrivalTime - b.ArrivalTime);

    const ordemExecucao = [];
    const metricas = [];
    let tempoAtual = 0;
    let somaEspera = 0;
    let somaTurnaround = 0;

    // Esta variável ajuda a saber se a CPU acabou de sair de um estado parada
    let veioDeOciosidade = true;

    for (let i = 0; i < processos.length; i++) {
        const p = processos[i];

        // Se o processo ainda não chegou, a CPU fica ociosa (parada)
        if (tempoAtual < p.ArrivalTime) {
            tempoAtual = p.ArrivalTime;
            veioDeOciosidade = true;
        }

        // Se NÃO veio de ociosidade e existe um tempo de troca de contexto (TTC)
        // Adicionamos esse tempo antes de começar o processo
        if (i > 0 && veioDeOciosidade === false && requisicao.TTC > 0) {
            ordemExecucao.push({
                PID: "TTC",
                Start: tempoAtual,
                End: tempoAtual + requisicao.TTC
            });
            tempoAtual = tempoAtual + requisicao.TTC;
        }

        // Agora o processo vai rodar
        veioDeOciosidade = false;
        const inicio = tempoAtual;
        tempoAtual = tempoAtual + p.BurstTime;
        const fim = tempoAtual;

        // Registramos a execução no gráfico de Gantt
        ordemExecucao.push({
            PID: p.PID,
            Start: inicio,
            End: fim
        });

        // Calculamos as métricas deste processo
        // Turnaround: quanto tempo o processo ficou no sistema (do início até o fim)
        const turnaround = fim - p.ArrivalTime;
        // Espera: tempo total no sistema menos o tempo que ele realmente rodou
        const espera = turnaround - p.BurstTime;

        metricas.push({
            PID: p.PID,
            EffectiveTime: turnaround,
            WaitingTime: espera
        });

        somaTurnaround += turnaround;
        somaEspera += espera;
    }

    // Calculamos as médias finais
    const totalProcessos = processos.length;
    return {
        ExecutionOrder: ordemExecucao,
        Metrics: metricas,
        AvgWaitTime: somaEspera / totalProcessos,
        AvgTurnaroundTime: somaTurnaround / totalProcessos
    };
}

module.exports = { runFCFS };